### PR TITLE
bump K8s to v1.25.16

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve?

Bump the K8s deps to v1.25.16.

Some APIs are deprecated (ref https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-25), but they don't affect TiDB Operator.

But the pod re-creation behavior of a Job has been changed in https://github.com/kubernetes/kubernetes/pull/110948, this affected one of our BR E2E cases

### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -28,9 +28,9 @@ require (
 	github.com/minio/minio-go/v6 v6.0.55
 	github.com/ncw/directio v1.0.5
 	github.com/onsi/ginkgo v1.14.1
-	github.com/onsi/gomega v1.10.2
-	github.com/openshift/generic-admission-server v1.14.1-0.20210422140326-da96454c926d
-	github.com/pingcap/advanced-statefulset/client v1.17.1-0.20231207074614-075934ad31d3
+	github.com/onsi/gomega v1.20.1
+	github.com/openshift/generic-admission-server v1.14.1-0.20221208154047-0cd6b8367d91
```
Comment (csuzhangxc) [inline go.mod:32]: ref https://github.com/openshift/generic-admission-server/pull/47

File: pkg/controller/dependences.go [hunk 2]
```diff
@@ -114,7 +114,7 @@ func DefaultCLIConfig() *CLIConfig {
 		LeaseDuration:          15 * time.Second,
 		RenewDeadline:          10 * time.Second,
 		RetryPeriod:            2 * time.Second,
-		ResourceLock:           resourcelock.EndpointsLeasesResourceLock,
+		ResourceLock:           resourcelock.LeasesResourceLock, // k8s uses leases by default from v1.20
```
Comment (csuzhangxc) [inline pkg/controller/dependences.go:117]: v1.6.0-alpha.10 used `EndpointsLeasesResourceLock`

File: tests/third_party/k8s/framework.go [hunk 3]
```diff
@@ -490,10 +489,6 @@ func (f *Framework) CreateNamespace(baseName string, labels map[string]string) (
 	// fail to create serviceAccount in it.
 	f.AddNamespacesToDelete(ns)
 
-	if err == nil && !f.SkipPrivilegedPSPBinding {
-		CreatePrivilegedPSPBinding(f.ClientSet, ns.Name)
-	}
```
Comment (csuzhangxc) [inline tests/third_party/k8s/framework.go:495]: Pod Security Policies was removed from Kubernetes in v1.25, ref https://kubernetes.io/docs/concepts/security/pod-security-policy/

File: hack/lib.sh [hunk 4]
```diff
@@ -24,15 +24,15 @@ OUTPUT=${ROOT}/output
...
 KUBECTL_BIN=$OUTPUT_BIN/kubectl
 HELM_BIN=$OUTPUT_BIN/helm
 DOCS_BIN=$OUTPUT_BIN/gen-crd-api-reference-docs
 CFSSL_BIN=$OUTPUT_BIN/cfssl
 CFSSLJSON_BIN=$OUTPUT_BIN/cfssljson
 JQ_BIN=$OUTPUT_BIN/jq
 CFSSL_VERSION=${CFSSL_VERSION:-1.2}
-K8S_VERSION=${K8S_VERSION:-0.24.17}
+K8S_VERSION=${K8S_VERSION:-0.25.16}
 JQ_VERSION=${JQ_VERSION:-1.6}
```
Comment (HuSharp) [inline hack/lib.sh:36]: There are three versions of this pr which are related k8s
- 0.25.16
- 1.25.16
- 1.25.9

Can you give me more information to help me check them? :)

Comment (csuzhangxc) [reply hack/lib.sh:36]: the target K8s version is v1.25.16 and it's sub mods named 0.25.16.

The v1.25.9 is the newest version of v1.25.x that KIND 0.19.0 can work.

Comment (HuSharp) [reply hack/lib.sh:36]: get it. thx!

**Review by HuSharp — APPROVED** [review]
LGTM

**Pull request merged**